libflux/src/semantic/walk.rs

```diff
@@ -0,0 +1,997 @@
+            Expression::Identifier(ref mut e) => Node::IdentifierExpr(e),
+            Expression::Array(ref mut e) => Node::ArrayExpr(e),
+            Expression::Function(ref mut e) => Node::FunctionExpr(e),
+            Expression::Logical(ref mut e) => Node::LogicalExpr(e),
+            Expression::Object(ref mut e) => Node::ObjectExpr(e),
+            Expression::Member(ref mut e) => Node::MemberExpr(e),
+            Expression::Index(ref mut e) => Node::IndexExpr(e),
+            Expression::Binary(ref mut e) => Node::BinaryExpr(e),
+            Expression::Unary(ref mut e) => Node::UnaryExpr(e),
+            Expression::Call(ref mut e) => Node::CallExpr(e),
+            Expression::Conditional(ref mut e) => Node::ConditionalExpr(e),
+            Expression::StringExpr(ref mut e) => Node::StringExpr(e),
+            Expression::Integer(ref mut e) => Node::IntegerLit(e),
+            Expression::Float(ref mut e) => Node::FloatLit(e),
+            Expression::StringLit(ref mut e) => Node::StringLit(e),
+            Expression::Duration(ref mut e) => Node::DurationLit(e),
+            Expression::Uint(ref mut e) => Node::UintLit(e),
+            Expression::Boolean(ref mut e) => Node::BooleanLit(e),
+            Expression::DateTime(ref mut e) => Node::DateTimeLit(e),
+            Expression::Regexp(ref mut e) => Node::RegexpLit(e),
+        }
+    }
+    fn from_stmt(stmt: &'a mut Statement) -> Node {
+        match *stmt {
+            Statement::Expr(ref mut s) => Node::ExprStmt(s),
+            Statement::Variable(ref mut s) => Node::VariableAssgn(s),
+            Statement::Option(ref mut s) => Node::OptionStmt(s),
+            Statement::Return(ref mut s) => Node::ReturnStmt(s),
+            Statement::Test(ref mut s) => Node::TestStmt(s),
+            Statement::Builtin(ref mut s) => Node::BuiltinStmt(s),
+        }
+    }
+    fn from_string_expr_part(sp: &'a mut StringExprPart) -> Node {
+        match *sp {
+            StringExprPart::Text(ref mut t) => Node::TextPart(t),
+            StringExprPart::Interpolated(ref mut e) => Node::InterpolatedPart(e),
+        }
+    }
+    fn from_assignment(a: &'a mut Assignment) -> Node {
+        match *a {
+            Assignment::Variable(ref mut v) => Node::VariableAssgn(v),
+            Assignment::Member(ref mut m) => Node::MemberAssgn(m),
+        }
+    }
+}
+
+// Visitor defines a visitor pattern for walking the AST
+//
+// When used with the walk function, Visit will be called for every node
+// in depth-first order. After all children for a Node have been visted,
+// Done is called on that Node to signal that we are done with that Node.
+//
+// If Visit returns None, walk will not recurse on the children.
+pub trait Visitor: Sized {
+    // Visit is called for a node.
+    // The returned visitor will be used to walk children of the node.
+    // If visit returns None, walk will not recurse on the children.
+    fn visit(&self, node: &mut Node) -> Option<Self>;
```

Comment:
Why does `visit` need a reference to `self` if it's returning another visitor? Why can't the signature be:
```Rust
fn visit(self, node: &mut Node) -> Option<Self>;
```

Reply:
I think it's better to let `visit` only borrows `self`.

This enables stuff like

```
let v = SomeVisitor{};
walk(&v, node);
println!("{}", v.something);
```

Otherwise, `visit` would own `v` and that couldn't be used after `walk`.

Reply:
I guess then a better question is why doesn't `walk` return a visitor? It seems a little weird that `visit` returns a visitor, but walk returns nothing?

Reply:
Mmmm ok, but what would you made of the _last_ visitor? wouldn't you need a "trace" of the visitors? Also, how can you access those visitors between calls?